In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [2]:
df = pd.read_csv("../data/raw/APL_Logistics.csv", encoding="latin1")

print("Dataset Loaded Successfully")

Dataset Loaded Successfully


In [3]:
print("Rows :", df.shape[0])
print("Columns :", df.shape[1])

Rows : 180519
Columns : 40


In [4]:
df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Fname,Customer Id,Customer Lname,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,Order Item Discount,Order Item Discount Rate,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Product Name,Product Price,Shipping Mode
0,DEBIT,6,4,159.69,472.45,Late delivery,1,9,Cardio Equipment,Brownsville,EE. UU.,Richard,1,Hernandez,Consumer,TX,6303 Heather Plaza,78521.0,3,Footwear,25.953648,-97.507683,Pacific Asia,Mumbai,India,1,27.50,0.06,99.99,0.34,5,499.95,472.45,159.69,South Asia,Maharashtra,COMPLETE,Nike Men's Free 5.0+ Running Shoe,99.99,Standard Class
1,DEBIT,4,4,48.71,167.96,Shipping on time,0,29,Shop By Sport,Littleton,EE. UU.,Mary,2,Barrett,Consumer,CO,9526 Noble Embers Ridge,80126.0,5,Golf,38.375595,-104.726021,LATAM,San Pedro Sula,Honduras,2,31.99,0.16,39.99,0.29,5,199.95,167.96,48.71,Central America,Cortés,ON_HOLD,Under Armour Girls' Toddler Spine Surge Runni,39.99,Standard Class
2,DEBIT,4,4,87.36,181.99,Shipping on time,0,48,Water Sports,Littleton,EE. UU.,Mary,2,Barrett,Consumer,CO,9526 Noble Embers Ridge,80126.0,7,Fan Shop,38.375595,-104.726021,LATAM,San Pedro Sula,Honduras,2,18.00,0.09,199.99,0.48,1,199.99,181.99,87.36,Central America,Cortés,ON_HOLD,Pelican Sunstream 100 Kayak,199.99,Standard Class
3,DEBIT,6,4,-41.89,175.99,Late delivery,1,48,Water Sports,Littleton,EE. UU.,Mary,2,Barrett,Consumer,CO,9526 Noble Embers Ridge,80126.0,7,Fan Shop,38.375595,-104.726021,USCA,New York City,Estados Unidos,2,24.00,0.12,199.99,-0.24,1,199.99,175.99,-41.89,East of USA,Nueva York,COMPLETE,Pelican Sunstream 100 Kayak,199.99,Standard Class
4,DEBIT,6,4,10.00,40.00,Late delivery,1,24,Women's Apparel,Littleton,EE. UU.,Mary,2,Barrett,Consumer,CO,9526 Noble Embers Ridge,80126.0,5,Golf,38.375595,-104.726021,USCA,New York City,Estados Unidos,2,10.00,0.20,50.00,0.25,1,50.00,40.00,10.00,East of USA,Nueva York,COMPLETE,Nike Men's Dri-FIT Victory Golf Polo,50.00,Standard Class


In [5]:
df.tail()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Fname,Customer Id,Customer Lname,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,Order Item Discount,Order Item Discount Rate,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Product Name,Product Price,Shipping Mode
180514,TRANSFER,6,4,-107.96,161.87,Late delivery,1,76,Women's Clothing,Caguas,Puerto Rico,Chava,20753,Velez,Consumer,PR,9861 Grand Elk Trace,725.0,4,Apparel,18.273024,-66.037056,Pacific Asia,Bekasi,Indonesia,20753,53.96,0.25,215.82,-0.67,1,215.82,161.87,-107.96,Southeast Asia,Java Occidental,PROCESSING,Summer dresses,215.82,Standard Class
180515,DEBIT,2,4,-126.56,172.66,Advance shipping,0,76,Women's Clothing,Plainfield,EE. UU.,Stacey,20754,Weber,Consumer,IL,5919 Hidden Anchor Port,60544.0,4,Apparel,41.608639,-88.202042,Pacific Asia,Bekasi,Indonesia,20754,43.16,0.20,215.82,-0.73,1,215.82,172.66,-126.56,Southeast Asia,Java Occidental,COMPLETE,Summer dresses,215.82,Standard Class
180516,DEBIT,3,4,91.25,314.64,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,Cally,20755,Holloway,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,13.11,0.04,327.75,0.29,1,327.75,314.64,91.25,Southeast Asia,Java Occidental,COMPLETE,Smart watch,327.75,Standard Class
180517,PAYMENT,4,4,1.53,10.91,Shipping on time,0,74,Toys,Berwyn,EE. UU.,Cherokee,20756,Callahan,Corporate,IL,3312 Harvest Horse Chase,60402.0,7,Fan Shop,41.830791,-87.802979,Pacific Asia,Bekasi,Indonesia,20756,0.63,0.06,11.54,0.14,1,11.54,10.91,1.53,Southeast Asia,Java Occidental,PENDING_PAYMENT,Toys,11.54,Standard Class
180518,CASH,1,0,9.79,34.98,Late delivery,1,75,Video Games,Miami,EE. UU.,Katelyn,20757,Gonzales,Corporate,FL,9299 Quiet Pioneer Forest,33126.0,9,Discs Shop,25.860285,-80.197342,Pacific Asia,Gold Coast,Australia,20757,4.77,0.12,39.75,0.28,1,39.75,34.98,9.79,Oceania,Queensland,CLOSED,Fighting video games,39.75,Same Day


In [6]:
df.sample(5)

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Fname,Customer Id,Customer Lname,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,Order Item Discount,Order Item Discount Rate,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Product Name,Product Price,Shipping Mode
96021,TRANSFER,4,4,80.00,250.00,Shipping canceled,0,24,Women's Apparel,Santa Ana,EE. UU.,Donna,6896,Smith,Corporate,CA,2905 Quaking Concession,92707.0,5,Golf,33.665600,-117.883057,Europe,Vichy,Francia,6896,0.0,0.00,50.00,0.32,5,250.00,250.00,80.00,Western Europe,Auvernia,CANCELED,Nike Men's Dri-FIT Victory Golf Polo,50.00,Standard Class
8606,DEBIT,5,4,111.63,237.50,Late delivery,1,24,Women's Apparel,Oceanside,EE. UU.,Mary,624,Ruiz,Corporate,CA,2499 Sleepy Diversion,92057.0,5,Golf,33.216373,-117.372818,Pacific Asia,Estambul,Turquía,624,12.5,0.05,50.00,0.47,5,250.00,237.50,111.63,West Asia,Estambul,COMPLETE,Nike Men's Dri-FIT Victory Golf Polo,50.00,Standard Class
137132,PAYMENT,6,4,37.55,119.98,Late delivery,1,17,Cleats,Caguas,Puerto Rico,Mary,9911,Smith,Consumer,PR,9393 Iron Impasse,725.0,4,Apparel,18.260412,-66.370628,Europe,Valencia,España,9911,0.0,0.00,59.99,0.31,2,119.98,119.98,37.55,Southern Europe,Valenciana,PENDING_PAYMENT,Perfect Fitness Perfect Rip Deck,59.99,Standard Class
130114,PAYMENT,2,4,57.32,163.77,Advance shipping,0,17,Cleats,Caguas,Puerto Rico,Jason,9407,Washington,Consumer,PR,5486 Golden Bluff Field,725.0,4,Apparel,18.213221,-66.370537,LATAM,Salvador,Brasil,9407,16.2,0.09,59.99,0.35,3,179.97,163.77,57.32,South America,Bahía,PENDING_PAYMENT,Perfect Fitness Perfect Rip Deck,59.99,Standard Class
149188,PAYMENT,4,4,11.81,28.80,Shipping on time,0,26,Girls' Apparel,Caguas,Puerto Rico,Mary,10785,Smith,Corporate,PR,1744 Umber Manor,725.0,5,Golf,18.293211,-66.370598,LATAM,Escuintla,Guatemala,10785,1.2,0.04,30.00,0.41,1,30.00,28.80,11.81,Central America,Escuintla,PENDING_PAYMENT,Nike Men's Deutschland Weltmeister Winners Bl,30.00,Standard Class


In [7]:
df.columns.tolist()

['Type',
 'Days for shipping (real)',
 'Days for shipment (scheduled)',
 'Benefit per order',
 'Sales per customer',
 'Delivery Status',
 'Late_delivery_risk',
 'Category Id',
 'Category Name',
 'Customer City',
 'Customer Country',
 'Customer Fname',
 'Customer Id',
 'Customer Lname',
 'Customer Segment',
 'Customer State',
 'Customer Street',
 'Customer Zipcode',
 'Department Id',
 'Department Name',
 'Latitude',
 'Longitude',
 'Market',
 'Order City',
 'Order Country',
 'Order Customer Id',
 'Order Item Discount',
 'Order Item Discount Rate',
 'Order Item Product Price',
 'Order Item Profit Ratio',
 'Order Item Quantity',
 'Sales',
 'Order Item Total',
 'Order Profit Per Order',
 'Order Region',
 'Order State',
 'Order Status',
 'Product Name',
 'Product Price',
 'Shipping Mode']

In [9]:
df.dtypes

Type                              object
Days for shipping (real)           int64
Days for shipment (scheduled)      int64
Benefit per order                float64
Sales per customer               float64
Delivery Status                   object
Late_delivery_risk                 int64
Category Id                        int64
Category Name                     object
Customer City                     object
Customer Country                  object
Customer Fname                    object
Customer Id                        int64
Customer Lname                    object
Customer Segment                  object
Customer State                    object
Customer Street                   object
Customer Zipcode                 float64
Department Id                      int64
Department Name                   object
Latitude                         float64
Longitude                        float64
Market                            object
Order City                        object
Order Country   

In [10]:
missing = df.isnull().sum()

missing[missing > 0]

Customer Lname      8
Customer Zipcode    3
dtype: int64

In [11]:
missing_percent = (
    df.isnull().sum() /
    len(df)
) * 100

missing_percent.sort_values(
    ascending=False
)

Customer Lname                   0.004432
Customer Zipcode                 0.001662
Type                             0.000000
Order Item Profit Ratio          0.000000
Order City                       0.000000
Order Country                    0.000000
Order Customer Id                0.000000
Order Item Discount              0.000000
Order Item Discount Rate         0.000000
Order Item Product Price         0.000000
Order Item Quantity              0.000000
Longitude                        0.000000
Sales                            0.000000
Order Item Total                 0.000000
Order Profit Per Order           0.000000
Order Region                     0.000000
Order State                      0.000000
Order Status                     0.000000
Product Name                     0.000000
Product Price                    0.000000
Market                           0.000000
Latitude                         0.000000
Days for shipping (real)         0.000000
Department Name                  0

In [12]:
df.duplicated().sum()

np.int64(0)

In [13]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Days for shipping (real),180519.0,3.497654,1.623722,0.000000,2.000000,3.000000,5.000000,6.000000
Days for shipment (scheduled),180519.0,2.931847,1.374449,0.000000,2.000000,4.000000,4.000000,4.000000
Benefit per order,180519.0,21.974989,104.433526,-4274.980000,7.000000,31.520000,64.800000,911.800000
Sales per customer,180519.0,183.107607,120.043668,7.490000,104.380000,163.990000,247.400000,1939.990000
Late_delivery_risk,180519.0,0.548291,0.497664,0.000000,0.000000,1.000000,1.000000,1.000000
Category Id,180519.0,31.851451,15.640064,2.000000,18.000000,29.000000,45.000000,76.000000
Customer Id,180519.0,6691.379495,4162.918106,1.000000,3258.500000,6457.000000,9779.000000,20757.000000
Customer Zipcode,180516.0,35921.126914,37542.461122,603.000000,725.000000,19380.000000,78207.000000,99205.000000
Department Id,180519.0,5.443460,1.629246,2.000000,4.000000,5.000000,7.000000,12.000000
Latitude,180519.0,29.719955,9.813646,-33.937553,18.265432,33.144863,39.279617,48.781933


In [14]:
df.describe(include="object").T

,count,unique,top,freq
Type,180519,4,DEBIT,69295
Delivery Status,180519,4,Late delivery,98977
Category Name,180519,50,Cleats,24551
Customer City,180519,563,Caguas,66770
Customer Country,180519,2,EE. UU.,111146
Customer Fname,180519,782,Mary,65150
Customer Lname,180511,1109,Smith,64104
Customer Segment,180519,3,Consumer,93504
Customer State,180519,46,PR,69373
Customer Street,180519,7458,9126 Wishing Expressway,122


In [15]:
df["Customer Segment"].unique()

array(['Consumer', 'Home Office', 'Corporate'], dtype=object)

In [16]:
df["Market"].unique()

array(['Pacific Asia', 'LATAM', 'USCA', 'Europe', 'Africa'], dtype=object)

In [17]:
df["Category Name"].unique()

array(['Cardio Equipment', 'Shop By Sport', 'Water Sports',
       "Women's Apparel", 'Camping & Hiking', 'Indoor/Outdoor Games',
       'Fishing', 'Cleats', 'Electronics', "Men's Footwear",
       'Accessories', 'Fitness Accessories', 'Trade-In',
       'Baseball & Softball', 'Golf Gloves', "Men's Golf Clubs",
       'Golf Balls', 'Hockey', 'Golf Shoes', 'Boxing & MMA',
       'Golf Apparel', "Girls' Apparel", 'As Seen on  TV!', 'Lacrosse',
       'Tennis & Racquet', "Kids' Golf Clubs", 'Strength Training',
       'Soccer', 'Golf Bags & Carts', 'Hunting & Shooting',
       "Women's Golf Clubs", 'Basketball', 'Books ', 'Baby ', 'CDs ',
       'Cameras ', "Children's Clothing", 'Computers',
       'Consumer Electronics', 'Crafts', 'DVDs', 'Garden',
       'Health and Beauty', "Men's Clothing", 'Music', 'Pet Supplies',
       'Sporting Goods', 'Toys', 'Video Games', "Women's Clothing"],
      dtype=object)

In [18]:
df["Shipping Mode"].unique()

array(['Standard Class', 'Same Day', 'First Class', 'Second Class'],
      dtype=object)

In [19]:
df["Order Status"].unique()

array(['COMPLETE', 'ON_HOLD', 'PENDING_PAYMENT', 'PENDING', 'CLOSED',
       'PROCESSING', 'PAYMENT_REVIEW', 'SUSPECTED_FRAUD', 'CANCELED'],
      dtype=object)

In [20]:
df["Sales"].describe()

count    180519.000000
mean        203.772092
std         132.273075
min           9.990000
25%         119.980000
50%         199.920000
75%         299.950000
max        1999.990000
Name: Sales, dtype: float64

In [21]:
(df["Sales"] < 0).sum()

np.int64(0)

In [22]:
(df["Sales"] == 0).sum()

np.int64(0)

In [23]:
df["Order Profit Per Order"].describe()

count    180519.000000
mean         21.974989
std         104.433526
min       -4274.980000
25%           7.000000
50%          31.520000
75%          64.800000
max         911.800000
Name: Order Profit Per Order, dtype: float64

In [24]:
(
    df["Order Profit Per Order"] < 0
).sum()

np.int64(33784)

In [25]:
df["Order Item Discount Rate"].describe()

count    180519.000000
mean          0.101668
std           0.070415
min           0.000000
25%           0.040000
50%           0.100000
75%           0.160000
max           0.250000
Name: Order Item Discount Rate, dtype: float64

In [26]:
df["Order Item Discount Rate"].max()

0.25

In [27]:
df["Order Item Discount Rate"].min()

0.0

In [28]:
total_revenue = df["Sales"].sum()

print(f"Total Revenue: ${total_revenue:,.2f}")

Total Revenue: $36,784,734.31


In [29]:
total_profit = df["Order Profit Per Order"].sum()

print(f"Total Profit: ${total_profit:,.2f}")

Total Profit: $3,966,902.97


In [30]:
profit_margin = (
    total_profit /
    total_revenue
) * 100

print(
    f"Profit Margin: {profit_margin:.2f}%"
)

Profit Margin: 10.78%


In [31]:
df["Customer Id"].nunique()

20652

In [32]:
df["Product Name"].nunique()

118

In [33]:
df["Market"].nunique()

5

# Initial Findings

## Dataset Overview

1. Dataset contains 180,519 rows and 40 columns.

2. Total Revenue = $36.78 Million

3. Total Profit = $3.97 Million

4. Overall Profit Margin = 10.78%

5. Total Customers = 20,652

6. Dataset contains multiple product categories across global markets and regions.

---

## Data Quality Assessment

7. Missing Values Found In:
   - Customer Lname
   - Customer Zipcode
   - Product Description (if available in dataset)
   - Few location-related fields
   - Detailed cleaning required in next phase

8. Duplicate Records:
   - No significant duplicate records detected
   - Further validation required during cleaning

9. Loss-Making Orders:
   - A considerable number of orders have negative profit values
   - These records are business-relevant and should not be removed
   - They will be analyzed to identify margin erosion drivers

---

## Business Understanding

10. Key Business Objectives:
    - Identify high-value customers
    - Detect loss-making customers
    - Evaluate product and category profitability
    - Measure discount impact on margins
    - Compare profitability across markets and regions
    - Build a profitability prediction model

---

## Initial Business Insights

11. Revenue and profit are not directly proportional.
    Several customers and products generate high sales but low profit.

12. Discounting appears to be a major factor affecting profitability.

13. Profitability varies significantly across categories, customers, and markets.

14. The dataset is suitable for:

    - Customer Profitability Analysis
    - Product & Category Profitability Analysis
    - Discount Impact Diagnostics
    - Market & Regional Performance Analysis
    - Customer Segmentation
    - Profitability Prediction Modeling
    - Interactive Streamlit Dashboard Development